Improve the porting guide page (#86)
OpenAMP/openamp-docs: docs/porting_guide.rst

Porting GuideLine
=================

********
Hardware
********

The porting of OpenAMP to a new :doc:`multicore system <../openamp/overview>` requires

> configuring the hardware on each core so it aligns with the OpenAMP architecture.

Reviewer: Proposal ```suggestion
The porting of OpenAMP to a new :doc:`multicore system <../openamp/overview>` requires
some  hardware for inter-processor communication.
```

Author: updated

This setup typically includes defining a shared memory region for
:ref:`RPMsg<overview-rpmsg-work-label>` based
:ref:`Interprocessor Communications (IPC)<ipc-work-label>`, with or without interrupts for
asynchronous inter-core notification and defining the firmware execution flow either independently
or with :ref:`Remoteproc<overview-Remoteproc-work-label>` and associated
:ref:`Resource Table<resource-table>`.

Memory and interrupt assignments are critical design choices for any port. For a broader overview, refer to

> :doc:`../protocol_details/system_considerations`.

Reviewer: From hardware perspective the OpenAMP require
- a shared memory for
  -  the RPmsg buffer 
  - the virtio rings
  - the resource table (optioal if the rpmsg host is not the Linux kernel)
-  an optional mailboxes ( burt recommanded ) acting a a ring bell interrupt on processors to inform about message reception

Author: I will use above and reword what is there.

Author: updated

Shared Memory
=============

Shared memory forms the :ref:`physical layer<rpmsg-layers-work-label>` for
:doc:`RPMsg <../docs/rpmsg_design>` protocol.
The specific memory type and layout are implementation dependent, but should be a dedicated

> SRAM or DDR region accessible by both cores, with caching disabled.

Reviewer: open-amp supports cache management

Author: updated

Memory requirements are generally modest because RPMsg is a control‑oriented protocol
rather than a high‑bandwidth streaming channel. For example, using the Linux RPMsg packet
size of 512 bytes, a 64kB shared memory region can hold roughly 128 messages — sufficient

> for most applications. Larger or smaller allocations can be chosen based on system needs.

Reviewer: you need also  memory for the virtio rings  , with Linux you should add 2* 4K for the vrings.. for other can be less

Author: added

Porting GuideLine
=================

********
Hardware
********

The porting of OpenAMP to a new :doc:`multicore system <../openamp/overview>` requires
configuring the hardware on each core so it aligns with the OpenAMP architecture.

This setup typically includes defining a shared memory region for
:ref:`RPMsg<overview-rpmsg-work-label>` based
:ref:`Interprocessor Communications (IPC)<ipc-work-label>`, with or without interrupts for
asynchronous inter-core notification and defining the firmware execution flow either independently
or with :ref:`Remoteproc<overview-remoteproc-work-label>` and associated
:ref:`Resource Table<resource-table>`.

Memory and interrupt assignments are critical design choices for any port. For a broader overview, refer to
:doc:`../protocol_details/system_considerations`.

Shared Memory
=============

Shared memory forms the :ref:`physical layer<rpmsg-layers-work-label>` for
:doc:`RPMsg <../docs/rpmsg_design>` protocol.
The specific memory type and layout are implementation dependent, but should be a dedicated

> SRAM or DDR region accessible by both cores, with caching disabled.

Reviewer: it work with or without cache. the cache management can be enabled in open-amp using the  `WITH_DCACHE` cmake configuration

Author: updated

Memory requirements are generally modest because :doc:`RPMsg <../docs/rpmsg_design>` is a
control‑oriented protocol rather than a high‑bandwidth streaming channel. For example, using
the Linux RPMsg packet size of 512 bytes, a 64kB shared memory region can hold roughly 128
messages — sufficient for most applications. Larger or smaller allocations can be chosen based
on system needs.

If the :ref:`Resource Table<resource-table>` is not embedded in the remote firmware image,
additional shared memory may be required for a dynamic table.

:ref:`Remoteproc<overview-remoteproc-work-label>` can also use shared memory for optional
trace buffers.

Memory Protection
-----------------

Because this is shared memory, appropriate hardware memory protection should be configured
on both processors.

Depending on the memory type, this may involve configuring the Memory Management Unit (MMU),
Memory Protection Unit (MPU), or Input-Output Memory Management Unit (IOMMU) to enforce

> correct access permissions.

Reviewer: Some specific mechanism are platform dependents, here is it a non exhaustive list:
 
Depending on the memory type, this may involve configuring some other  protection units such as  the Memory Management Unit (MMU), Memory Protection Unit (MPU),   or Input-Output Memory Management Unit (IOMMU) .

Author: updated

On systems running an advanced OS — such as Linux on the main
processor — these protections may be applied through OS mechanisms like the device tree or
via the :ref:`Remoteproc<overview-remoteproc-work-label>` :ref:`Resource Table<resource-table>`.

Notification
------------

:doc:`RPMsg <../docs/rpmsg_design>` uses :ref:`ring buffers<rpmsg-protocol-mac>` in shared
memory, so either processor can poll for incoming messages. However, asynchronous notification
via interrupts is recommended.

Most heterogeneous SoCs include a built‑in inter‑core interrupt mechanism, often called a
mailbox. These implementations typically combine shared memory with interrupt signaling and

> may be managed through an Inter‑Processor Communication Controller (IPCC).

Reviewer: IPCC is not a common terminology and can lead to confusinon as it is the name of the hardware mailbox for the stm32MP Socs.
i would remove this part of the sentence. 
another point is that some mailboxes does not embed shared memory

`Most heterogeneous SoCs include a built‑in inter‑core notification peripheral, often called a mailbox. These implementations typically combine interrupt signaling with optional internal shared memory . `

Author: updated

If no dedicated hardware is available — or it is reserved for other purposes —

> software‑generated interrupts can be used instead.

Reviewer: I do not understand this sentence. In my view, the statement is incorrect. It is not possible to generate an inter-core software interrupt. Could you clarify your meaning?

Author: was not aware of this either, added because of mention in ticket
"List common implementations (hardware mailbox, software interrupt, etc.)."
can remove if its not possible.

Reviewer: > was not aware of this either, added because of mention in ticket "List common implementations (hardware mailbox, software interrupt, etc.)." can remove if its not possible.

the request is to give few detail about the mailbox usage in IPC. for instance:

"The mailbox is used for inter-processor communication to notify the cores, through a software interrupt, that a vring has been updated for message reception, buffer release, or both.

The core can then manage the message in normal context, for example, in a thread, or in interrupt context."

Author: updated

***************
Porting Options
***************

OpenAMP consists of two major components: :ref:`Remoteproc<overview-remoteproc-work-label>`
and :doc:`RPMsg <../docs/rpmsg_design>`. These can be ported

> independently or together, at either the driver level or the device level.

Reviewer: ```suggestion
independently or together, at either the virtio driver  or the virtio device role.
```

Driver level ports integrate with an operating system’s existing frameworks,
while device level ports implement the functionality directly on bare metal without

> leveraging OS‑provided drivers.

Reviewer: The sentence is not correct and need to be reworked.

- the driver/device notion only concern the RPMsg, not the remoteproc
- The main processor usually implements the driver  while the remote processor implements the device role
- Both roles can be implemented with or without OS. the only exception is the Linux driver that only implement the driver role.

Author: I think we have a case of terminology overload throwing me there. I took driver as anything not baremetal. Will need to rethink how to phrase it according to your points, maybe as part of chapter reconstruction.

Reviewer: > I think we have a case of terminology overload throwing me there. I took driver as anything not baremetal. Will need to rethink how to phrase it according to your points, maybe as part of chapter reconstruction.

Interesting feedback demonstrating that this notion of driver/device need to be clarify in the doc. 😉

`libmetal <https://github.com/OpenAMP/libmetal>`_ provides the
:ref:`hardware abstraction layer<hardware-abstraction>` for both.

The main porting approaches include:

- :ref:`Remoteproc<overview-remoteproc-work-label>` on the remote processor only,
  with the main processor using an existing
  :ref:`Remoteproc<overview-remoteproc-work-label>` implementation (e.g., Linux Remoteproc)

>   and no IPC.

Reviewer: it is the inverse. the remoteproc is usually implemented on the main processor. its main feature is to load a firmware on the remote processor and manage its livecycle
the Linux kernel implement its own remoteproc framework

Author: Similarly I think of "remoteproc" as all things remoteproc, including any remote side parts needed, like I presume the resource table is, but could be way off.

Reviewer: One criticism of OpenAMP is that the remote processor framework is required on the remote device side. that why I would split remoteproc in 3 feature:
- remote processor live cycle management ( implemented on main processor)
- remoteproc virtio : the transport layer of the virtio protocol
- the resource table 

In fact, on the remote side, the full remote processor framework is not required. Only the remote processor virtio transport layer is required.  the resource table is also only mandatory for communication with Linux.
In addition, if no inter-process communication (IPC) is implemented, the remote side does not need the remote processor framework.

see my proposal in comment https://github.com/OpenAMP/openamp-docs/pull/86/changes#r3340380433

- :doc:`RPMsg <../docs/rpmsg_design>` on the remote processor only, with the main processor
  using an existing :doc:`RPMsg <../docs/rpmsg_design>` stack (e.g., Linux RPMsg) and no
  remote firmware management.

- Custom device‑level implementation of :ref:`Remoteproc<overview-remoteproc-work-label>`

>   and/or :doc:`RPMsg <../docs/rpmsg_design>` for both processors.

Reviewer: suggestion 

```
The main porting approaches include:
- on main processor
  - :ref:`Remoteproc<overview-remoteproc-work-label>`  driver role to manage remoteprocessor Live Cycle
  -  :ref :`remoteproc virtio <TBD>`  to implement the virtio driver remoteproc transport
  -   :doc:`RPMsg <../docs/rpmsg_design>` driver role
transport layer
- on remote processor
  - optional :ref:`Remoteproc<overview-remoteproc-work-label>` device role for the resource table management
  -  :ref :`remoteproc virtio <TBD>`  device role to manage the virtio device remoteproc transport
  -  :doc:`RPMsg <../docs/rpmsg_design>` device role

Custom device‑level implementation of [Remoteproc](https://openamp--86.org.readthedocs.build/en/86/openamp/overview.html#overview-remoteproc-work-label) and/or [RPMsg](https://openamp--86.org.readthedocs.build/en/86/docs/rpmsg_design.html) for both processors.
```

.. _driver-lcm-remoteproc:

> Driver Lifecycle Management via Remoteproc

Reviewer: ```suggestion
Remote processor Lifecycle Management via Remoteproc
```

Reviewer: not updated?

==========================================

> Some systems do not require IPC or use an alternative IPC mechanism. In these cases, only

Reviewer: to remove the remoteproc framework is independent from the rpmsg

Author: not quite sure here that what is written is opposite to your point. will re-think and re-visit as part of rewrite

Reviewer: if no IPC no remoteproc needed on remote processors

:ref:`Remoteproc<overview-remoteproc-work-label>` may be ported (or reused, as on Linux)

> on both the main and remote processors.

Reviewer: the remoteproc framework is optional
- need on main processor if it need to manage the remote processor live cyccle
- can be used on remoteprocessor to manage the resource table for IPC

Author: ditto to previous

:ref:`Remoteproc<overview-remoteproc-work-label>` and :doc:`RPMsg <../docs/rpmsg_design>` may
be required without any driver‑level abstraction.
In this case, the full :ref:`Remoteproc<overview-remoteproc-work-label>` and
:doc:`RPMsg <../docs/rpmsg_design>` mechanism must be implemented
directly on both the main and remote processors, ensuring that each core provides the necessary
firmware lifecycle management, messaging, shared‑memory handling, and notification logic without
relying on OS‑level drivers or frameworks.

- Pros: Lightweight.
- Cons: Highly custom and less portable.

.. _device-rpmsg:

Device Level Custom RPMsg
=========================

In a simpler bare‑metal only environments, a port of only :doc:`RPMsg <../docs/rpmsg_design>` may
be required without any driver‑level abstraction.
In this case, only the :doc:`RPMsg <../docs/rpmsg_design>` mechanism need be implemented
directly on both the main and remote processors, ensuring that each core provides the necessary
IPC and notification logic without relying on OS‑level drivers or frameworks.

- Pros: Very lightweight.
- Cons: Highly custom and less portable.

> 

Reviewer: I wonder if we should not restructure the chapters.. here is a proposal  (from .. `_driver-lcm-remoteproc: reference`)
i just put titles and list content 

Don't hesitate to challenge this , it look to me that we woill probably need few iterations with some other reviewers to land to a good porting guide 

```
Remoteproc framework 
====================
- remote processor management
- resource table management

driver role 
---------------
- remoteprocessor livecycle management
- resource table discovery
-  example:  https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw

device role 
---------------
- optional 
- can be use for the resource table management 
    example: https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/legacy_apps/machine/xlnx/zynqmp_r5/platform_info.c
- alternative is to managea static resource table directly
    example https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c

Remoteproc virtio framework 
========================
This is the virtio transport layer  ( virtio MMIO is another)

driver role 
---------------
 -  get resource table information and configure it
 -  create virtio driver 
 
device role 
---------------
 -  get resource table information filled by the virtio driver
 -  create virtio device 
    example https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c 

RPMsg framework 
===============
- IPC 
- based on VirtIO
   - remoteproc virtio
   -  static vring transport layer 
       example:https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/ipc/openamp 

reference to https://github.com/OpenAMP/openamp-system-reference examples for the driver and devices implementation

```

Author: ok will revisit and see.

Author: Addressed the simpler items, but for the more major ones not sure at this point exactly how to proceed. Also wondering if some of the suggestion belong in the porting page or should they be in a design page as we might be crossing boundaries between porting and implementation options?

Reviewer: In my view, for this pages users will likely approach this with a requirement to implement either the driver or the device topology. With this approach, our role would be to guide them in identifying the appropriate topology (driver or device) and then selecting the assets to integrate, depending on the topology, the presence of Linux on main processor and the required features.

Author: Tried to restructure based on your suggested headings above.

    - the :ref:`Resource Table<resource-table>`
      (optional if the RPMsg host is not the Linux kernel)

- Optional, but recommended, mailboxes acting as a ring bell interrupt on processors to
  notify of message reception

Memory and interrupt assignments are critical design choices for any port. For a broader overview, refer to
:doc:`../protocol_details/system_considerations`.

Shared Memory
=============

Shared memory forms the :ref:`physical layer<rpmsg-layers-work-label>` for
:doc:`RPMsg <../docs/rpmsg_design>` protocol.
The specific memory type and layout are implementation dependent, but should be a dedicated
SRAM or DDR region accessible by both cores. Caching is enabled in OpenAMP with the
`WITH_DCACHE <https://github.com/OpenAMP/open-amp/blob/main/cmake/options.cmake>`_
cmake configuration.

Memory requirements are generally modest because :doc:`RPMsg <../docs/rpmsg_design>` is a
control‑oriented protocol rather than a high‑bandwidth streaming channel. For example, using
the Linux RPMsg packet size of 512 bytes, a 64kB shared memory region can hold roughly 128
messages — sufficient for most applications. Larger or smaller allocations can be chosen based

> on system needs.

Reviewer: 64 KB is very big for some systems . limit to 16 messages  -> 8 KB

"Larger or smaller allocations can be chosen based
on system needs"
-> explain instead that the size and the number of messgae can be tuned based on the system needs and constraints

In addition to the messages, the Virtio ring structures need memory allocated. For Linux

> this equates to 2*4k but likely less for other implementations.

Reviewer: for linux the size of the allocated vring is aligned on a linux kernel Page ( usually 4K)

If the :ref:`Resource Table<resource-table>` is not embedded in the remote firmware image,
additional shared memory may be required for a dynamic table.

:ref:`Remoteproc<overview-remoteproc-work-label>` can also use shared memory for optional
trace buffers.

Memory Protection
-----------------

Because this is shared memory, appropriate hardware memory protection should be configured
on both processors.

Depending on the memory type, this may involve configuring the Memory Management Unit (MMU),
Memory Protection Unit (MPU), or Input-Output Memory Management Unit (IOMMU) to enforce
correct access permissions.

On systems running an advanced OS — such as Linux on the main
processor — these protections may be applied through OS mechanisms like the device tree or

> via the :ref:`Remoteproc<overview-remoteproc-work-label>` :ref:`Resource Table<resource-table>`.

Reviewer: Confusing, I would remove this sentence. protections can not be applied trough resource table and are platform dependent. this can be done by some other mechanisms

on both the main and remote processors.

The main processor uses driver level :ref:`Remoteproc<overview-remoteproc-work-label>` to
load, start, stop, and manage remote firmware.
This approach is useful when the remote firmware must be externally controlled or when
multiple firmware images may be deployed depending on runtime needs.
This configuration is common in custom or bare‑metal remote environments.

- Pros: Full remote firmware management
- Cons: No IPC. Larger software footprint

.. _driver-lcm-remoteproc-rpmsg:

Driver Lifecycle Management via Remoteproc with IPC
===================================================

This setup is the most complete and includes driver based lifecycle management and IPC.
In these cases, both :ref:`Remoteproc<overview-remoteproc-work-label>` and
:doc:`RPMsg <../docs/rpmsg_design>` need to be ported (or reused, as on Linux) on both
the main and remote processors.

The main processor uses driver level :ref:`Remoteproc<overview-remoteproc-work-label>` and
:doc:`RPMsg <../docs/rpmsg_design>`.

This approach is useful when full IPC and remote firmware control is required.

This configuration is common in custom or bare‑metal remote environments.

> - Pros: Full IPC and remote firmware management

Reviewer: This is one main point that is not necessarily true for the remote processor side. We try to minimize the footprint on the remote side even in this topology.
For me, there is usually a mix-up between the remoteproc virtio, which is the virtio transport layer, and the remoteproc that manages the load and lifecycle of the remote processor. 

i would like try to have a more explicit split for the remote proc framework.

------------

:doc:`RPMsg <../docs/rpmsg_design>` uses :ref:`ring buffers<rpmsg-protocol-mac>` in shared
memory, so either processor can poll for incoming messages. However, asynchronous notification
via interrupts is recommended.

Most heterogeneous SoCs include a built‑in inter‑core interrupt mechanism, often called a
mailbox. The sending core raises the mailbox interrupt to notify the other core that a vring has
been updated for message reception, buffer release, or both.
The core can then manage the message in normal context, for example, in a thread, or in
interrupt.

***************
Porting Options
***************

OpenAMP consists of two major components: :ref:`Remoteproc<overview-remoteproc-work-label>`
and :doc:`RPMsg <../docs/rpmsg_design>`. These can be ported
independently or together.

For both there is a concept of a driver and device role which affect the virtio transport layer,
specified when setting up the virtio structure, via the definitions:

:openamp_doc_link:`VIRTIO_DEV_DEVICE <VIRTIO_DEV_DEVICE>` or
:openamp_doc_link:`VIRTIO_DEV_DRIVER <VIRTIO_DEV_DRIVER>`

When creating the virtio device (not to be confused with the device role), the driver role

> indicates the main processor and the device role the remote processor.

Reviewer 2: This `the driver role indicates the main processor and the device role the remote processor` is always true with Linux as host, but OpenAMP permits a remote to take the driver role and some Zephyr configurations do exactly that.
So, here, I suggest to add `in the common Linux-host case...`.

Porting GuideLine
=================

********
Hardware
********

The porting of OpenAMP to a new :doc:`multicore system <../openamp/overview>` requires
some hardware for inter-processor communication.

- Shared memory for

    - the :ref:`RPMsg<overview-rpmsg-work-label>` buffer
    - the :ref:`Virtio Rings<docs/data_structures_content:Shared virtqueue structure>`
    - the :ref:`Resource Table<resource-table>`

>       (optional if the RPMsg host is not the Linux kernel)

Reviewer 3: Is the resource table really optional if RPMsg host is not the linux?

I think if host and remote both is using open-amp library for RPMsg communication then we need resource table.

@arnopo should we remove this optional comment? Or we if you are aware of any such example, we should provide an example for such a case?

Reviewer: @tnmysh : here are 2 zephyr samples without  resource table:
- https://github.com/zephyrproject-rtos/zephyr/tree/main/subsys/ipc/open-amp
- https://github.com/zephyrproject-rtos/zephyr/tree/main/subsys/ipc/ipc_service

- Optional, but recommended, mailboxes acting as a ring bell interrupt on processors to
  notify of message reception

Memory and interrupt assignments are critical design choices for any port. For a broader overview, refer to
:doc:`../protocol_details/system_considerations`.

Shared Memory
=============

Shared memory forms the :ref:`physical layer<rpmsg-layers-work-label>` for
:doc:`RPMsg <../docs/rpmsg_design>` protocol.
The specific memory type and layout are implementation dependent, but should be a dedicated

> SRAM or DDR region accessible by both cores. Caching is enabled in OpenAMP with the

Reviewer 4: Whether or not the memory is caching memory is not chosen by OpenAMP, it is an intrinsic property of the memory as defined by the host operating system. The `WITH_DCACHE` flag only sets if OpenAMP should do cache maintenance operations or not.

Additionally, since the cacheability of a memory region is a property of the hardware/port we have agreed to move this flag to libmetal where the other hardware specific options are located. This whole last sentence can just be dropped.

Reviewer: > Additionally, since the cacheability of a memory region is a property of the hardware/port we have agreed to move this flag to libmetal where the other hardware specific options are located. This whole last sentence can just be dropped.

I agree with the objective. However, since it has not been done yet, we need to document the current implementation in the porting guide.

Reviewer 4: Sure, but this sentence is not matching what we have currently either. That flag doesn't set if caching is enabled or not, only if we perform cache ops or not, suggest:

"Cache maintenance operations are performed by OpenAMP if the `WITH_DCACHE` cmake flag is set. It should be enabled if your shared memory region is caching and the core you are building for is not cache coherent with the remote side."

Reviewer 3: I agree with @gleno on improving the sentence with some changes.

%s/OpenAMP/open-amp/ We refer to OpenAMP as a whole project, and open-amp as a library. I would replace 'OpenAMP' with open-amp and insert link to the open-amp library repo.

Another question, should we also mention DMA operations can be used, and zero-copy is possible?

@arnopo I am not aware on the support of zero-copy in the open-amp library. If it is available, then it is worth mentioning.

Author: Replaced sentence.
@tnmysh I think currently the main distinction in the documentation is "OpenAMP" for the libary and "OpenAMP Project" for the project. I will effect other pages too if it is to change.

Reviewer: > @arnopo I am not aware on the support of zero-copy in the open-amp library. If it is available, then it is worth mentioning.

good point I have created an issue : https://github.com/OpenAMP/openamp-docs/issues/95

    life cycle
  - :openamp_doc_link:`Remoteproc Virtio <remoteproc_virtio>` to implement the virtio driver transport layer
  - :openamp_doc_link:`RPMsg <rpmsg_device>`

- On the remote processor

  - Optional :openamp_doc_link:`Remoteproc <remoteproc>` as device role for the
    :openamp_doc_link:`resource table management <resource_table>`
  - :openamp_doc_link:`Remoteproc Virtio <remoteproc_virtio>` device role to implement the
    virtio device transport layer
  - :openamp_doc_link:`RPMsg <rpmsg_device>`

Additional details and examples are provided in the subsequent sections.

Remoteproc Framework
====================

The Remoteproc framework as the name suggests is responsible for remote processor management.
For both roles this includes resource table management and/or setup of the virtio transport
and as the driver role also the remote processor lifecycle management.

Remoteproc Driver Role
----------------------

In the driver role, Remoteproc includes  management of the remote processor lifecycle and
resource table discovery.

Most :ref:`OpenAMP Samples and Demos<demos-reference-samples>` provide examples of lifecycle management for Remoteproc
in the driver role.

The

> `load_fw <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw>`_

Reviewer 3: I am not sure, if we should mention `load_fw` demo. It's not being verified for a very long time, and not sure if it is even working.

I vote to remove this line. @arnopo and others have any opinion on this?

Reviewer 2: I haven't tested this.
If we intend to keep it, we should test it for every release. Otherwise, I'd vote to remove it.

Reviewer: I am quite puzzled by this. It has the advantage of providing an application example. If we decide not to document it, we should also remove it from the system reference repository.

is an example dedicated to demonstrating Remoteproc lifecycle management.

Remoteproc Device Role
----------------------

Use of Remoteproc framework for the remote or device role is optional and when used is for

> resource table management, dynamically or statically.

Reviewer 3: We should define what 'dynamically' and 'statically' means?

Author: adjusted

Remoteproc Virtio Framework
===========================

When using Remoteproc, the
`Virtio transport layer <https://github.com/OpenAMP/open-amp/main/lib/remoteproc/remoteproc_virtio.c>`_
is created through it (as compared to through
`RPMsg's Virtio <https://github.com/OpenAMP/open-amp/blob/main/lib/rpmsg/rpmsg_virtio.c>`_).

Remoteproc Virtio Driver Role
-----------------------------

For the main processor driver role, Remoteproc is used

- to obtain and configure resource table information.
- create the main processor or driver side Virtio device

The intent of OpenAMP Remoteproc is to remain compatible with Linux's Remoteproc.

The
`Multi Services Example <https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services>`_
example uses Linux Remoteproc for the driver role implementation.
For details refer to :ref:`RPMsg Multi Services<rpmsg-multi-services-intro>` Demo.

Remoteproc Virtio Device Role
-----------------------------

For the remote processor device role, Remoteproc is used

- to get resource table information populated by the Virtio driver
- create remote or device side Virtio device

The

> `Multi Services <https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c>`_

Reviewer 4: I've been wanting to zap this example for a bit now, there is an example in Zephyr upstream that does the exact same thing plus more. We should point to it here and above instead so we don't need to update these docs later. `https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c`

Reviewer: There is a common part, such as rpmsg_tty and the RP client thread, but the OpenAMP example also demonstrates:

- the raw character device
- how to create/release RPMsg channel from the Linux side
- how to create fixed endpoints

This sample could be rewritten, for instance by splitting it into several smaller samples, but not simply removed.

Reviewer 4: > how to create/release RPMsg channel from the Linux side

Does it? Looks like it just points to the upstream Linux examples and gives some text on using it from Linux, which the Zephyr upstream example does also.

What is being documented here doesn't include the raw char dev nor fixed endpoints, only creating multiple services and using virtio in the device configuration. Both of which are done by the example in Zephyr upstream.

Reviewer: if you look at https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/README.rst

- Demos 1 & 2  can be reproduced with the Zephyr openamp_rsc_table sample.
- Demo 3 is the creation of a channel from Linux with an NS announcement on the Zephyr side. This requires code that is not supported in the Zephyr openamp_rsc_table sample.
- Demos 4 and 5 use RPMsg raw, which is not supported in the Zephyr openamp_rsc_table sample.

Reviewer 4: I'd question the usefulness of raw messages in general, and if we do think they have value then we should upstream the same to the Zephyr openamp_rsc_table sample.

Anyway, this is a debate better had over on the issue tracker [0] if I want to push for it again. For now since the demo still does exist in the system-reference then no issue with referencing it in documentation for now.

[0] https://github.com/OpenAMP/openamp-system-reference/issues/41

Porting GuideLine
=================

********
Hardware
********

The porting of OpenAMP to a new :doc:`multicore system <../openamp/overview>` requires
some hardware for inter-processor communication.

- Shared memory for

    - the :ref:`RPMsg<overview-rpmsg-work-label>` buffer
    - the :ref:`Virtio Rings<docs/data_structures_content:Shared virtqueue structure>`
    - the :ref:`Resource Table<resource-table>`
      (optional if the RPMsg host is not the Linux kernel)

> - Optional, but recommended, mailboxes acting as a ring bell interrupt on processors to

Reviewer 2: s/`ring bell interrupt`/`doorbell interrupt`

  notify of message reception

Memory and interrupt assignments are critical design choices for any port. For a broader overview, refer to
:doc:`../protocol_details/system_considerations`.

Shared Memory
=============

Shared memory forms the :ref:`physical layer<rpmsg-layers-work-label>` for
:doc:`RPMsg <../docs/rpmsg_design>` protocol.
The specific memory type and layout are implementation dependent, but should be a dedicated
SRAM or DDR region accessible by both cores.
Cache maintenance operations are performed by OpenAMP if the

> `WITH_DCACHE <https://github.com/OpenAMP/open-amp/blob/main/cmake/options.cmake>`_ cmake flag

Reviewer 2: open-amp exposes finer-grained options: `WITH_DCACHE_VRINGS`, `WITH_DCACHE_BUFFERS`, `WITH_DCACHE_RSC_TABLE`.
On systems where only part of the shared region is cacheable, the granular flags matter for both correctness and cycle cost.

Reviewer: @iuliana-prodan : The use of these cmake options has been deprecated (by you) in v2023.10 :smiley: . The code has been removed for now

Reviewer 2: You're right, @arnopo, it's here: [OpenAMP commit.](https://github.com/OpenAMP/open-amp/commit/f2162a69ac6764749a2e326d580921095c2527d9)

I completely forgot about that :(

> is set. It should be enabled if your shared memory region is caching and the core you are

Reviewer 2: s/` shared memory region is caching`/ `shared memory region is cacheable`

building for is not cache coherent with the remote side.

Memory requirements are generally modest because :doc:`RPMsg <../docs/rpmsg_design>` is a
control‑oriented protocol rather than a high‑bandwidth streaming channel. For example, using
the Linux RPMsg packet size of 512 bytes, a 8kB shared memory region can hold roughly 16

> messages. The size and message count should be tuned based on system needs.

Reviewer 2: Shared-memory sizing math is misleading. Linux’s `virtio_rpmsg_bus` splits the buffer pool half TX / half RX, so 8 kB gives 8 buffers per direction, not 16 messages.
The 512 bytes includes the 16-byte `struct rpmsg_hdr`, so usable payload is 496 bytes.

I suggest rephasing: `For example, with Linux’s 512-byte RPMsg buffer size (496 bytes of payload after the 16-byte header), an 8 kB buffer region provides 16 buffers - 8 per direction, since the pool is split between TX and RX. This is in addition to the vring memory described below.`

mailbox. The sending core raises the mailbox interrupt to notify the other core that a vring has
been updated for message reception, buffer release, or both.
The core can then manage the message in normal context, for example, in a thread, or in
interrupt.

***************
Porting Options
***************

OpenAMP consists of two major components: :ref:`Remoteproc<overview-remoteproc-work-label>`
and :doc:`RPMsg <../docs/rpmsg_design>`. These can be ported
independently or together.

For both there is a concept of a driver and device role which affect the virtio transport layer,
specified when setting up the virtio structure, via the definitions:

:openamp_doc_link:`VIRTIO_DEV_DEVICE <VIRTIO_DEV_DEVICE>` or
:openamp_doc_link:`VIRTIO_DEV_DRIVER <VIRTIO_DEV_DRIVER>`

When creating the virtio device (not to be confused with the device role), the driver role
indicates the main processor and the device role the remote processor.

In OpenAMP
`RPMsg Virtio <https://github.com/OpenAMP/open-amp/blob/main/lib/include/openamp/rpmsg_virtio.h>`_
these are aliased as
:openamp_doc_link:`RPMSG_HOST <RPMSG_HOST>` and :openamp_doc_link:`RPMSG_REMOTE <RPMSG_REMOTE>`

> respectively.

Reviewer 2: `Respectively` binds `VIRTIO_DEV_DEVICE` to `RPMSG_HOST` and `VIRTIO_DEV_DRIVER` to `RPMSG_REMOTE`. That is backwards - `rpmsg_virtio.h` defines `RPMSG_HOST` as `VIRTIO_DEV_DRIVER` and `RPMSG_REMOTE` as `VIRTIO_DEV_DEVICE`.

resource table discovery.

Most :ref:`OpenAMP Samples and Demos<demos-reference-samples>` provide examples of lifecycle management for Remoteproc
in the driver role.

The
`load_fw <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw>`_
is an example dedicated to demonstrating Remoteproc lifecycle management.

Remoteproc Device Role
----------------------

Use of Remoteproc framework for the remote or device role is optional and when used is for
resource table management, parsing the resource table for dynamic resource allocation or
statically assigning the resources required for Remoteproc.

Dynamic examples are:

- `ZynqMP R5 <https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/legacy_apps/machine/xlnx/zynqmp_r5/platform_info.c>`_
  machine implementation example.

Management using a static resource examples are:

- `Multi Services Remote Example <https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c>`_.

Remoteproc Virtio Framework
===========================

When using Remoteproc, the

> `Virtio transport layer <https://github.com/OpenAMP/open-amp/main/lib/remoteproc/remoteproc_virtio.c>`_

Reviewer 2: ```suggestion
`Virtio transport layer <https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c>`_
```

Broken link, missing `blob` segment.

From `load_fw.c <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw/load_fw.c>`_

.. literalinclude::  ../openamp-system-reference/examples/legacy_apps/examples/load_fw/load_fw.c
   :language: c
   :lines: 21-57

.. _port-rpmsg:

***********************
Platform Specific RPMsg
***********************

In OpenAMP, :doc:`RPMsg <../docs/rpmsg_design>` uses
`VirtIO <https://docs.oasis-open.org/virtio/virtio/>`_ to manage shared buffers.
The OpenAMP library provides a
`Remoteproc VirtIO <https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c>`_
backend implementation, and a
`VirtIO and RPMsg <https://github.com/OpenAMP/open-amp/blob/main/lib/rpmsg/rpmsg_virtio.c>`_
implementation.

You can also implement your own VirtIO backend using the
`OpenAMP VirtIO <https://github.com/OpenAMP/open-amp/tree/main/lib/virtio>`_ and
`RPMsg <https://github.com/OpenAMP/open-amp/tree/main/lib/rpmsg>`_ components provided by OpenAMP.

If you choose to create your own backend, you can use the
`Remoteproc VirtIO <https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c>`_
backend as a reference.

> For an example of setting up Remoteproc and Virtio for a remote device refer to

Reviewer 3: "Setting up remoteproc and virtio for a remote device'.

I think it's worth mentioning that to setup 'remoteproc' ops, you'd need platform driver: For example: Cortex-R5 remoteproc ops are defined here: https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/legacy_apps/machine/xlnx/zynqmp_r5/zynqmp_r5_a53_rproc.c

Reviewer 3: I am sorry, reference to `zynqmp_r5_a53_rproc.c` should go in the section: `platform specific remoteproc driver`.

Then in the section `platform sepcific rpmsg` we should rewrite the line as:

"Setting up remoteproc virtio for a remote device'.

The 'and' gives idea that we are talking about 'remoteproc' separately, but we are really talking about 'remoteproc virtio'. 

Thanks.

Author: added example to platform specific remoteproc driver and
remove the and from remoteproc virtio

> `zynqmp platform_info.c <https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/legacy_apps/machine/xlnx/zynqmp/platform_info.c>`_

Reviewer 3: %s/zynqmp platform_info.c/zynqmp_r5 platform_info.c

and Please use this link instead: https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/legacy_apps/machine/xlnx/zynqmp_r5/platform_info.c

Author: changed

=========================================

With the :ref:`remoteproc driver functions<port-remoteproc-driver>` required
by the framework ported, the user can call the :ref:`Remoteproc<overview-remoteproc-work-label>`
APIs to run an application on a remote processor, as described in the
:ref:`Remote User APIs<remoteproc_config>` section of the Remoteproc design.

The following code snippets from the
`Load FW System Reference Example <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw>`_
demonstrate the use of the Remote User APIs.

Remoteproc Init
---------------

From `platform_info.c <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw/platform_info.c>`_

.. literalinclude::  ../openamp-system-reference/examples/legacy_apps/examples/load_fw/platform_info.c
   :language: c
   :lines: 16-31

Lifecycle APIs
--------------

From `load_fw.c <https://github.com/OpenAMP/openamp-system-reference/tree/main/examples/legacy_apps/examples/load_fw/load_fw.c>`_

.. literalinclude::  ../openamp-system-reference/examples/legacy_apps/examples/load_fw/load_fw.c
   :language: c

>    :lines: 21-57

Reviewer 2: Instead of line numbers which may need adjusting each release, try:
```
.. literalinclude:: ../openamp-system-reference/examples/legacy_apps/examples/load_fw/load_fw.c
   :language: c
   :start-at: int load_exectuable_block
   :end-before: /* end of lifecycle demo */
```

.. _port-rpmsg:

***********************
Platform Specific RPMsg
***********************

In OpenAMP, :doc:`RPMsg <../docs/rpmsg_design>` uses
`VirtIO <https://docs.oasis-open.org/virtio/virtio/>`_ to manage shared buffers.
The OpenAMP library provides a
`Remoteproc VirtIO <https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c>`_
backend implementation, and a
`VirtIO and RPMsg <https://github.com/OpenAMP/open-amp/blob/main/lib/rpmsg/rpmsg_virtio.c>`_
implementation.

You can also implement your own VirtIO backend using the
`OpenAMP VirtIO <https://github.com/OpenAMP/open-amp/tree/main/lib/virtio>`_ and
`RPMsg <https://github.com/OpenAMP/open-amp/tree/main/lib/rpmsg>`_ components provided by OpenAMP.

If you choose to create your own backend, you can use the
`Remoteproc VirtIO <https://github.com/OpenAMP/open-amp/blob/main/lib/remoteproc/remoteproc_virtio.c>`_
backend as a reference.

> For an example of setting up Remoteproc Virtio for a remote device refer to

Reviewer 2: `Virtio` vs `VirtIO` or `Remoteproc` vs `remoteproc` vs `Remote Proc Design`.
Pick one of each and enforce it.
